src/moonlink_backend/src/error.rs

```diff
@@ -1,89 +1,159 @@
-    #[error("Moonlink metadata error: {source}")]
-    MoonlinkMetadataStoreError { source: MoonlinkMetadataStoreError },
+    #[error("{0}")]
+    ParseIntError(ErrorStruct),
+
+    #[error("{0}")]
+    PostgresSource(ErrorStruct),
+
+    #[error("{0}")]
+    Io(ErrorStruct),
+
+    #[error("{0}")]
+    MoonlinkConnectorError(ErrorStruct),
+
+    #[error("{0}")]
+    MoonlinkError(ErrorStruct),
+
+    #[error("{0}")]
+    MoonlinkMetadataStoreError(ErrorStruct),
+
     #[error("Invalid argument: {0}")]
     InvalidArgumentError(String),
-    #[error("Watch channel receive error: {source}")]
-    TokioWatchRecvError {
-        source: tokio::sync::watch::error::RecvError,
-    },
-    #[error("JSON serialization/deserialization error: {source}")]
-    Json { source: Arc<serde_json::Error> },
+
+    #[error("{0}")]
+    TokioWatchRecvError(ErrorStruct),
+    #[error("{0}")]
+    Json(ErrorStruct),
 }
 
 pub type Result<T> = result::Result<T, Error>;
 
 impl From<PostgresSourceError> for Error {
+    #[track_caller]
     fn from(source: PostgresSourceError) -> Self {
-        Error::PostgresSource {
-            source: Arc::new(source),
-        }
+        Error::PostgresSource(ErrorStruct {
+            message: format!("Postgres source error: {source}"),
+            status: ErrorStatus::Temporary,
+            source: Some(Arc::new(source.into())),
+            location: Some(Location::caller()),
+        })
     }
 }
 
 impl From<ParseIntError> for Error {
+    #[track_caller]
     fn from(source: ParseIntError) -> Self {
-        Error::ParseIntError { source }
+        Error::ParseIntError(ErrorStruct {
+            message: format!("Parse integer error: {source}"),
+            status: ErrorStatus::Permanent,
+            source: Some(Arc::new(source.into())),
+            location: Some(Location::caller()),
+        })
     }
 }
 
 impl From<MoonlinkConnectorError> for Error {
+    #[track_caller]
     fn from(source: MoonlinkConnectorError) -> Self {
-        Error::MoonlinkConnectorError { source }
+        Error::MoonlinkConnectorError(ErrorStruct {
+            message: format!("Moonlink connector error: {source}"),
+            status: ErrorStatus::Temporary,
+            source: Some(Arc::new(source.into())),
+            location: Some(Location::caller()),
+        })
     }
 }
 
 impl From<MoonlinkError> for Error {
+    #[track_caller]
     fn from(source: MoonlinkError) -> Self {
-        Error::MoonlinkError { source }
+        Error::MoonlinkError(ErrorStruct {
+            message: format!("Moonlink error: {source}"),
+            status: ErrorStatus::Temporary,
+            source: Some(Arc::new(source.into())),
+            location: Some(Location::caller()),
+        })
     }
 }
 
 impl From<MoonlinkMetadataStoreError> for Error {
+    #[track_caller]
     fn from(source: MoonlinkMetadataStoreError) -> Self {
-        Error::MoonlinkMetadataStoreError { source }
+        Error::MoonlinkMetadataStoreError(ErrorStruct {
```

Comment:
I think we need to use enum and check whether each sub-error catagory is of temporary / permanent error